### PR TITLE
Fix 5993 - Remove Usage of Shared Materials with Clipping Primitives

### PR DESCRIPTION
**Pull request by Cameron-Micka**
## Overview
This change fixes an issue where clipping primitives were modifying a renderers shared material in the editor. Materials are now cloned (much like how renderer.material works at run time). This was done to support edit time manipulation without leaking materials into the scene.

https://github.com/microsoft/MixedRealityToolkit-Unity/pull/6001 relies on this change.

## Changes
- Fixes: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5993

## Verification
Please try using a scene that uses primitive clipping such as the ClippingExamples or MaterialGallery scenes.

>
> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

### CODE REVIEW

File: Assets/MixedRealityToolkit/Inspectors/ClippingPrimitiveInspector.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+using Microsoft.MixedReality.Toolkit.Utilities;
+using System.Linq;
+using UnityEditor;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Editor
+{
+    [CustomEditor(typeof(ClippingPrimitive))]
+    public abstract class ClippingPrimitiveEditor : UnityEditor.Editor
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Inspectors/ClippingPrimitiveInspector.cs]: please comment public class

**Comment by Cameron-Micka** [comment]
@julenka of course! This PR shouldn't change any performance characteristics (clipping primitives still have to create material clones to change shader keywords, like clipping primitives did before this change). The public API also remains the same, so hopefully no breaking changes.

The biggest difference is that in edit mode the original system would call [renderer.sharedMaterial](https://docs.unity3d.com/ScriptReference/Renderer-sharedMaterial.html) rather than [renderer.material](https://docs.unity3d.com/ScriptReference/Renderer-material.html) because calling renderer.material at edit time emits this Unity error: 

> Instantiating material due to calling renderer.material during edit mode. This will leak materials into the scene. You most likely want to use renderer.sharedMaterial instead.

But, using renderer.sharedMaterial will affect all instances that use that material (I thought at edit time all shared materials were unique but I was wrong about that).

So, now I always call renderer.sharedMaterial (at edit and run time) and create a material clone to assign. This keeps the code path the same at edit time and run time since renderer.material does a similar thing. From the renderer.material docs: 

> This function automatically instantiates the materials and makes them unique to this renderer. It is your responsibility to destroy the materials when the game object is being destroyed. Resources.UnloadUnusedAssets also destroys the materials but it is usually only called when loading a new level.

The material clone's lifetime gets managed now and destroyed when the clipping primitive is destroyed and when a renderer is removed.

I've seen other projects write a generic material referencing system to avoid some of this bespoke logic. But, that would be a somewhat substantial change to the MRTK (maybe for anther day?). Also, we can luckily lean on material property blocks more and more these days, so that helps too!

The TL;DR is this change should fix situations where global materials were erroneously modified while making sure all cloned created materials are tracked and destoryed, and not accidently serialized into a scene.

File: Assets/MixedRealityToolkit/Inspectors/ClippingPrimitiveInspector.cs (hunk 1 above)
Comment (Cameron-Micka) [reply Assets/MixedRealityToolkit/Inspectors/ClippingPrimitiveInspector.cs]: Nice catch, fixed!

File: Assets/MixedRealityToolkit/Inspectors/ClippingPrimitiveInspector.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        public override void OnInspectorGUI()
+        {
+            var clippingPrimitive = (ClippingPrimitive)target;
+
+            var previousRenderers = clippingPrimitive.GetRenderersCopy();
+            DrawDefaultInspector();
+            var currentRenderers = clippingPrimitive.GetRenderersCopy();
+
+            // Add or remove and renderers that were added or removed via the inspector.
+            foreach (var renderer in previousRenderers.Except(currentRenderers))
```
Comment (julenka) [inline Assets/MixedRealityToolkit/Inspectors/ClippingPrimitiveInspector.cs]: 😎
